[visionOS] Audio-only playback does not appear to originate from its hosted window

jernoble opened the pull request:
#### 778e409cfeb9f52381f185896296c2d14b3cceed
<pre>
[visionOS] Audio-only playback does not appear to originate from its hosted window
<a href="https://bugs.webkit.org/show_bug.cgi?id=271492">https://bugs.webkit.org/show_bug.cgi?id=271492</a>
<a href="https://rdar.apple.com/125259675">rdar://125259675</a>

Reviewed by Andy Estes.

Introduce the concept of a &quot;default spatial tracking label&quot; by creating a label in the UI process
and adding it to a new UIView child of WKContentView. That label is passed down to the WebContent
process and GPU process, where it is consumed by MediaPlayerPrivate objects in the GPU process.

When the page is visible, and the MediaPlayerPrivate has no video representation, this new &quot;default&quot;
label is used. Otherwise, allow the platform to determine the label itself, and assign it to both
the video and audio presentations automatically.

When the page is _not_ visible, use the AVAudioSession&apos;s spatial tracking label.

* Source/WebCore/PAL/pal/spi/cocoa/AVFoundationSPI.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::defaultSpatialTrackingLabel const):
(WebCore::Document::defaultSpatialTrackingLabelChanged):
(WebCore::Document::addDefaultSpatialTrackingLabelChangedObserver):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::m_defaultSpatialTrackingLabelChangedObserver):
(WebCore::HTMLMediaElement::visibilityStateChanged):
(WebCore::HTMLMediaElement::updateSpatialTrackingLabel):
(WebCore::HTMLMediaElement::spatialTrackingLabel const):
(WebCore::HTMLMediaElement::setSpatialTrackingLabel):
(WebCore::HTMLMediaElement::defaultSpatialTrackingLabelChanged):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setDefaultSpatialTrackingLabel):
* Source/WebCore/page/Page.h:
(WebCore::Page::defaultSpatialTrackingLabel const):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::loadWithNextMediaEngine):
(WebCore::MediaPlayer::defaultSpatialTrackingLabel const):
(WebCore::MediaPlayer::setDefaultSpatialTrackingLabel):
(WebCore::MediaPlayer::spatialTrackingLabel const):
(WebCore::MediaPlayer::setSpatialTrackingLabel):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::defaultSpatialTrackingLabel const):
(WebCore::MediaPlayerPrivateInterface::setDefaultSpatialTrackingLabel):
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
(WebCore::MediaPlayerPrivateAVFoundation::isVisible const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::MediaPlayerPrivateAVFoundationObjC):
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVPlayer):
(WebCore::MediaPlayerPrivateAVFoundationObjC::platformSetVisible):
(WebCore::MediaPlayerPrivateAVFoundationObjC::defaultSpatialTrackingLabel const):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setDefaultSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateAVFoundationObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::MediaPlayerPrivateMediaSourceAVFObjC):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::defaultSpatialTrackingLabel const):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setDefaultSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setDefaultSpatialTrackingLabel):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _fullScreenPlaceholderView]):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::setSpatialTrackingLabel):
(WebKit::PlaybackSessionManagerProxy::setSpatialTrackingLabel):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setSpatialTrackingLabel):
(WebKit::WebPageProxy::spatialTrackingLabel const):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::spatialTrackingLabel const):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
(-[WKContentView spatialTrackingLabel]):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::defaultSpatialTrackingLabel const):
(WebKit::MediaPlayerPrivateRemote::setDefaultSpatialTrackingLabel):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setDefaultSpatialTrackingLabel):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::setSpatialTrackingLabel):

Canonical link: <a href="https://commits.webkit.org/276971@main">https://commits.webkit.org/276971@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5731414755c35df02ccfb69916bf770a180be923

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45882 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25010 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48553 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41920 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48188 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29305 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22408 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37542 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46460 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22082 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39582 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18718 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19487 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40677 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3926 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42181 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41027 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50323 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20876 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17370 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44678 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22178 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43558 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10247 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22537 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21869 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/PAL/pal/spi/cocoa/AVFoundationSPI.h:
```diff
@@ -486,3 +486,9 @@ NS_ASSUME_NONNULL_END
 #endif
+
+#if PLATFORM(VISION)
+@interface AVSampleBufferAudioRenderer (AVSampleBufferVideoRendererWebKitPrivate)
```
`AVSampleBufferAudioRendererWebKitPrivate` might be more appropriate

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -9328,18 +9339,41 @@ String HTMLMediaElement::localizedSourceType() const
+{
+    if (m_spatialTrackingLabel == spatialTrackingLabel)
+        return;
+    m_spatialTrackingLabel = spatialTrackingLabel;
```
`m_spatialTrackingLabel = WTFMove(spatialTrackingLabel)`

eric-carlson commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4744,4 +4744,17 @@ bool Page::hasActiveImmersiveSession() const
 #endif // PLATFORM(IOS_FAMILY) && ENABLE(WEBXR)
 
+#if HAVE(SPATIAL_TRACKING_LABEL)
+void Page::setDefaultSpatialTrackingLabel(const String& label)
```
Can this be `String&&`?

aestes commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7952,6 +7952,30 @@ void Document::addDisplayChangedObserver(const DisplayChangedObserver& observer)
+#if HAVE(SPATIAL_TRACKING_LABEL)
+const String& Document::defaultSpatialTrackingLabel() const
+{
+    if (page())
```
Per our refcounting guidelines:
```suggestion
    if (RefPtr page = protectedPage())
```

eric-carlson commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13805,6 +13805,18 @@ void WebPageProxy::updateRemoteFrameAccessibilityOffset(WebCore::FrameIdentifier
 }
 
+#if HAVE(SPATIAL_TRACKING_LABEL)
+void WebPageProxy::setSpatialTrackingLabel(const String& spatialTrackingLabel)
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9407,6 +9407,13 @@ void WebPage::remoteViewPointToRootView(FrameIdentifier frameID, FloatPoint poin
 }
 
+#if HAVE(SPATIAL_TRACKING_LABEL)
+void WebPage::setDefaultSpatialTrackingLabel(const String& label)
```
Ditto

aestes commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7952,6 +7952,30 @@ void Document::addDisplayChangedObserver(const DisplayChangedObserver& observer)
+{
+    for (auto& observer : copyToVector(m_defaultSpatialTrackingLabelChangedObservers)) {
+        if (observer)
+            (*observer)(defaultSpatialTrackingLabel.isolatedCopy());
```
curious why we have to `isolatedCopy` here.

jernoble replied:
This is removed in a more recent version of this patch; I just missed this code in the `String&& -> const String&` conversion.

aestes commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -9328,18 +9339,41 @@ String HTMLMediaElement::localizedSourceType() const
+
+    m_player->setSpatialTrackingLabel(String { m_spatialTrackingLabel });
+
+    auto page = document().page();
```
```suggestion
    RefPtr page = document().protectedPage();
```

aestes commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4744,4 +4744,17 @@ bool Page::hasActiveImmersiveSession() const
+{
+    if (m_defaultSpatialTrackingLabel == label)
+        return;
+    m_defaultSpatialTrackingLabel = WTFMove(label);
```
```suggestion
    m_defaultSpatialTrackingLabel = label;
```

aestes commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -497,6 +497,9 @@ HTMLMediaElement::HTMLMediaElement(const QualifiedName& tagName, Document& docum
     , m_modeAtMostRecentPlayback(AudioSessionMode::Default)
 #endif
+#if HAVE(SPATIAL_TRACKING_LABEL)
+    , m_defaultSpatialTrackingLabelChangedObserver([this] (String&& defaultSpatialTrackingLabel) { defaultSpatialTrackingLabelChanged(WTFMove(defaultSpatialTrackingLabel)); })
```
Should this be a `const String&`?

aestes commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -620,6 +620,10 @@ void MediaPlayer::loadWithNextMediaEngine(const MediaPlayerFactory* current)
 #endif
             m_private->prepareForPlayback(m_inPrivateBrowsingMode, m_preload, m_preservesPitch, m_shouldPrepareToPlay, m_shouldPrepareToRender);
+#if HAVE(SPATIAL_TRACKING_LABEL)
+            m_private->setDefaultSpatialTrackingLabel(String { m_defaultSpatialTrackingLabel });
```
Do you need to construct a new `String` here?

jernoble replied:
Nope!

aestes commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -620,6 +620,10 @@ void MediaPlayer::loadWithNextMediaEngine(const MediaPlayerFactory* current)
             m_private->prepareForPlayback(m_inPrivateBrowsingMode, m_preload, m_preservesPitch, m_shouldPrepareToPlay, m_shouldPrepareToRender);
+#if HAVE(SPATIAL_TRACKING_LABEL)
+            m_private->setDefaultSpatialTrackingLabel(String { m_defaultSpatialTrackingLabel });
+            m_private->setSpatialTrackingLabel(String { m_spatialTrackingLabel });
```
Ditto

aestes commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -1938,15 +1942,33 @@ void MediaPlayer::setShouldCheckHardwareSupport(bool value)
+    if (m_defaultSpatialTrackingLabel == defaultSpatialTrackingLabel)
+        return;
+    m_defaultSpatialTrackingLabel = defaultSpatialTrackingLabel;
+    m_private->setDefaultSpatialTrackingLabel(WTFMove(defaultSpatialTrackingLabel));
```
```suggestion
    m_private->setDefaultSpatialTrackingLabel(defaultSpatialTrackingLabel);
```

aestes commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -1938,15 +1942,33 @@ void MediaPlayer::setShouldCheckHardwareSupport(bool value)
+    if (m_spatialTrackingLabel == spatialTrackingLabel)
+        return;
+    m_spatialTrackingLabel = spatialTrackingLabel;
     m_private->setSpatialTrackingLabel(WTFMove(spatialTrackingLabel));
```
```suggestion
    m_private->setSpatialTrackingLabel(spatialTrackingLabel);
```

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -4025,18 +4037,63 @@ void determineChangedTracksFromNewTracksAndOldItems(MediaSelectionGroupAVFObjC*
+{
+    if (m_defaultSpatialTrackingLabel == defaultSpatialTrackingLabel)
+        return;
+    m_defaultSpatialTrackingLabel = WTFMove(defaultSpatialTrackingLabel);
```
```suggestion
    m_defaultSpatialTrackingLabel = defaultSpatialTrackingLabel;
```

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -4025,18 +4037,63 @@ void determineChangedTracksFromNewTracksAndOldItems(MediaSelectionGroupAVFObjC*
 {
     if (m_spatialTrackingLabel == spatialTrackingLabel)
         return;
     m_spatialTrackingLabel = WTFMove(spatialTrackingLabel);
```
```suggestion
    m_spatialTrackingLabel = spatialTrackingLabel;
```

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -4025,18 +4037,63 @@ void determineChangedTracksFromNewTracksAndOldItems(MediaSelectionGroupAVFObjC*
+    // If there is no AVPlayerLayer, use the default spatial tracking label if available, or
+    // the session's spatial tracking label if not.
+    if (!m_defaultSpatialTrackingLabel.isNull())
+        [m_avPlayer _setSTSLabel:nsStringNilIfNull(m_defaultSpatialTrackingLabel)];
```
No need for `nsStringNilIfNull` if we're already checking for null on the line above.

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -1675,11 +1693,36 @@ void getSupportedTypes(HashSet<String>& types) const final
+    // If there is no video renderer, use the default spatial tracking label if available, or
+    // the session's spatial tracking label if not, and set the label directly on each audio
+    // renderer.
+    auto *defaultLabel = !m_defaultSpatialTrackingLabel.isNull() ? (NSString *)m_defaultSpatialTrackingLabel : session.spatialTrackingLabel;
```
Nit: If you declared this as `NSString *defaultLabel` do you still need the `(NSString *)` cast?

jernoble replied:
Yes, it complains because the two halves of the ternary operator don't have matching types.

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -1675,11 +1693,36 @@ void getSupportedTypes(HashSet<String>& types) const final
+    // the session's spatial tracking label if not, and set the label directly on each audio
+    // renderer.
+    auto *defaultLabel = !m_defaultSpatialTrackingLabel.isNull() ? (NSString *)m_defaultSpatialTrackingLabel : session.spatialTrackingLabel;
+    for (const auto &key : m_sampleBufferAudioRendererMap.keys())
```
```suggestion
    for (const auto& key : m_sampleBufferAudioRendererMap.keys())
```

jernoble replied:
This is an Objective-C type, so coding style guidelines say this should go on the right.

aestes replied:
Well, it's a reference to a pointer; I don't think the fact that it refers to an Objective-C type changes our guidelines for reference placement (but I don't feel strongly!)

aestes commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
```diff
@@ -1231,10 +1231,17 @@ void RemoteMediaPlayerProxy::setShouldCheckHardwareSupport(bool value)
+#if HAVE(SPATIAL_TRACKING_LABEL)
+void RemoteMediaPlayerProxy::setDefaultSpatialTrackingLabel(const String& defaultSpatialTrackingLabel)
+{
+    m_player->setDefaultSpatialTrackingLabel(WTFMove(defaultSpatialTrackingLabel));
```
```suggestion
    m_player->setDefaultSpatialTrackingLabel(defaultSpatialTrackingLabel);
```

aestes commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -263,6 +269,15 @@ - (instancetype)_commonInitializationWithProcessPool:(WebKit::WebProcessPool&)pr
+    _spatialTrackingLabel = makeString("WKContentView Label: "_s, createVersion4UUIDString());
+    [[_spatialTrackingView layer] setValue:(NSString *)_spatialTrackingLabel forKeyPath:@"separatedOptions.STSLabel"];
+    [self addSubview:_spatialTrackingView.get()];
+    _page->setSpatialTrackingLabel(_spatialTrackingLabel);
```
Will this result in audio emitting from the top-leading corner of the content view?

jernoble replied:
Good question! I'll make sure this gets the equivalent of `width:100%; height:100%`.

aestes replied:
Or zero-sized but centered.

aestes commented on Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
```diff
@@ -1731,19 +1731,35 @@ void MediaPlayerPrivateRemote::setShouldCheckHardwareSupport(bool value)
+    if (defaultSpatialTrackingLabel == m_defaultSpatialTrackingLabel)
+        return;
+
+    m_defaultSpatialTrackingLabel = WTFMove(defaultSpatialTrackingLabel);
```
```suggestion
    m_defaultSpatialTrackingLabel = defaultSpatialTrackingLabel;
```

aestes commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9428,6 +9428,13 @@ void WebPage::adjustVisibilityForTargetedElements(const Vector<std::pair<WebCore
+#if HAVE(SPATIAL_TRACKING_LABEL)
+void WebPage::setDefaultSpatialTrackingLabel(const String& label)
+{
+    corePage()->setDefaultSpatialTrackingLabel(WTFMove(label));
```
I think `Page` needs to be protected here.

jernoble committed:
[visionOS] Audio-only playback does not appear to originate from its hosted window
https://bugs.webkit.org/show_bug.cgi?id=271492
rdar://125259675

Reviewed by Andy Estes.

Introduce the concept of a "default spatial tracking label" by creating a label in the UI process
and adding it to a new UIView child of WKContentView. That label is passed down to the WebContent
process and GPU process, where it is consumed by MediaPlayerPrivate objects in the GPU process.

When the page is visible, and the MediaPlayerPrivate has no video representation, this new "default"
label is used. Otherwise, allow the platform to determine the label itself, and assign it to both
the video and audio presentations automatically.

When the page is _not_ visible, use the AVAudioSession's spatial tracking label.

* Source/WebCore/PAL/pal/spi/cocoa/AVFoundationSPI.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::defaultSpatialTrackingLabel const):
(WebCore::Document::defaultSpatialTrackingLabelChanged):
(WebCore::Document::addDefaultSpatialTrackingLabelChangedObserver):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::m_defaultSpatialTrackingLabelChangedObserver):
(WebCore::HTMLMediaElement::visibilityStateChanged):
(WebCore::HTMLMediaElement::updateSpatialTrackingLabel):
(WebCore::HTMLMediaElement::spatialTrackingLabel const):
(WebCore::HTMLMediaElement::setSpatialTrackingLabel):
(WebCore::HTMLMediaElement::defaultSpatialTrackingLabelChanged):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setDefaultSpatialTrackingLabel):
* Source/WebCore/page/Page.h:
(WebCore::Page::defaultSpatialTrackingLabel const):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::loadWithNextMediaEngine):
(WebCore::MediaPlayer::defaultSpatialTrackingLabel const):
(WebCore::MediaPlayer::setDefaultSpatialTrackingLabel):
(WebCore::MediaPlayer::spatialTrackingLabel const):
(WebCore::MediaPlayer::setSpatialTrackingLabel):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::defaultSpatialTrackingLabel const):
(WebCore::MediaPlayerPrivateInterface::setDefaultSpatialTrackingLabel):
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
(WebCore::MediaPlayerPrivateAVFoundation::isVisible const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::MediaPlayerPrivateAVFoundationObjC):
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVPlayer):
(WebCore::MediaPlayerPrivateAVFoundationObjC::platformSetVisible):
(WebCore::MediaPlayerPrivateAVFoundationObjC::defaultSpatialTrackingLabel const):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setDefaultSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateAVFoundationObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::MediaPlayerPrivateMediaSourceAVFObjC):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::defaultSpatialTrackingLabel const):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setDefaultSpatialTrackingLabel):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setDefaultSpatialTrackingLabel):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _fullScreenPlaceholderView]):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::setSpatialTrackingLabel):
(WebKit::PlaybackSessionManagerProxy::setSpatialTrackingLabel):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setSpatialTrackingLabel):
(WebKit::WebPageProxy::spatialTrackingLabel const):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::spatialTrackingLabel const):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _commonInitializationWithProcessPool:configuration:]):
(-[WKContentView spatialTrackingLabel]):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::defaultSpatialTrackingLabel const):
(WebKit::MediaPlayerPrivateRemote::setDefaultSpatialTrackingLabel):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setDefaultSpatialTrackingLabel):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::setSpatialTrackingLabel):

Canonical link: https://commits.webkit.org/276971@main
